### PR TITLE
VZ-5962 Generalisation of BOM Validation Process

### PR DESCRIPTION
**Pull request by absBitm**
# Description
we rectified the following issues with the bom validation process.
1.  Now cluster images / tags will be matched to vz bom, earlier it was in reverse order.
2. Added feature of allowed or whitelisted namespaces list
3. Cleaned up code redundancies
4. Simplified code of many unnecessary nested loops
5. Improvised data structures, replaced static ds to dynamic  

Fixes [VZ-5962](https://jira.oraclecorp.com/jira/browse/VZ-5962)

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: tools/bom-validator/bom-validator.go [hunk 1]
```diff
@@ -72,43 +70,49 @@ type knownIssues struct {
...
-	"rancher-webhook": {alternateTags: []string{"v0.1.1", "v0.1.2", "v0.1.4"}, message: rancherWarningMessage},
-	"fleet-agent":     {alternateTags: []string{"v0.3.5"}, message: rancherWarningMessage},
-	"fleet":           {alternateTags: []string{"v0.3.5"}, message: rancherWarningMessage},
-	"gitjob":          {alternateTags: []string{"v0.1.15"}, message: rancherWarningMessage},
-	"oraclelinux":     {alternateTags: []string{"7-slim", "7.9"}, message: oracleLinuxWarningMessage},
+	"rancher-webhook":              {alternateTags: []string{"v0.1.1", "v0.1.2", "v0.1.4"}, message: rancherWarningMessage},
+	"fleet-agent":                  {alternateTags: []string{"v0.3.5"}, message: rancherWarningMessage},
+	"fleet":                        {alternateTags: []string{"v0.3.5"}, message: rancherWarningMessage},
+	"gitjob":                       {alternateTags: []string{"v0.1.15"}, message: rancherWarningMessage},
+	"shell":                        {alternateTags: []string{"v0.1.6"}, message: rancherWarningMessage},
```
Comment (michael-cico) [inline tools/bom-validator/bom-validator.go:77]: new one?  I haven't seen `shell` pop up before.

Comment (absBitm) [reply tools/bom-validator/bom-validator.go:77]: Yes it came in both the upgrade pipelines (admin cluster), 
https://build.verrazzano.io/blue/organizations/jenkins/verrazzano-upgrade-tests/detail/vicky%2Fbom-generalisations/1/pipeline
https://build.verrazzano.io/blue/organizations/jenkins/verrazzano-upgrade-path-tests/detail/vicky%2Fbom-generalisations/6/pipeline

File: tools/bom-validator/bom-validator.go [hunk 2]
```diff
@@ -72,43 +70,49 @@ type knownIssues struct {
...
+	"example-helidon-greet-app-v1": {alternateTags: []string{"1.0.0-1-20210728181814-eb1e622"}, message: imageMissingMessage},
```
[the remark this answers is not in the document]
Comment (absBitm) [reply tools/bom-validator/bom-validator.go:78]: correct sir,
I will do the same

File: tools/bom-validator/bom-validator.go [hunk 3]
```diff
@@ -157,152 +161,131 @@ func getBOM(vBom *verrazzanoBom) {
...
 			fmt.Printf("Warning: Image Name = %s: %s\n", name, msg)
 		}
 	}
-	fmt.Println()
-	// Dump Images that don't match BOM, Failure
-	if !passedValidation {
-		fmt.Println("Image Errors: BOM Images that don't match Cluster Images")
+	if len(clusterImagesNotFound) > 0 {
+		fmt.Println()
+		fmt.Println("Image Errors: CLuster Images not mentioned into bom")
```
Comment (michael-cico) [inline tools/bom-validator/bom-validator.go:238]: Suggested error message text:

```
Image Errors: Images found in Verrazzano namespaces in cluster not declared in BOM
```

File: tools/bom-validator/bom-validator.go [hunk 4]
```diff
@@ -157,152 +161,131 @@ func getBOM(vBom *verrazzanoBom) {
...
+		fmt.Println(textDivider)
+		for name, tag := range clusterImagesNotFound {
+			fmt.Printf("Image not mentioned into bom: %s:%s\n", name, tag)
+		}
+		error = true
+	}
+	if len(clusterImageTagErrors) > 0 {
+		fmt.Println()
+		fmt.Println("Image's Tag Errors: Cluster Image's Tags doesn't mentioned into Bom Image's Tags")
```
Comment (michael-cico) [inline tools/bom-validator/bom-validator.go:247]: Suggested error message text:

```
Image Errors: Images found in cluster with unexpected tags
```

File: tools/bom-validator/bom-validator.go (hunk 3 above)
Comment (absBitm) [reply tools/bom-validator/bom-validator.go:238]: I guess we should remove "verrazzano namespace"
as we are validating beyond verrazzano namespace 

is this fine?
Image Errors: Images found in allowed namespaces in cluster not declared in BOM

File: tools/bom-validator/bom-validator.go [hunk 5]
```diff
@@ -157,152 +161,131 @@ func getBOM(vBom *verrazzanoBom) {
...
-	fmt.Println()
-	// Dump Images that don't match BOM, Failure
-	if !passedValidation {
-		fmt.Println("Image Errors: BOM Images that don't match Cluster Images")
+	if len(clusterImagesNotFound) > 0 {
+		fmt.Println()
+		fmt.Println("Image Errors: CLuster Images not mentioned into bom")
+		fmt.Println(textDivider)
+		for name, tag := range clusterImagesNotFound {
+			fmt.Printf("Image not mentioned into bom: %s:%s\n", name, tag)
```
Comment (michael-cico) [inline tools/bom-validator/bom-validator.go:241]: Suggested error text for the format string:

```
Found image in allowed namespace not declared in BOM: %s:%s\n
```

File: tools/bom-validator/bom-validator.go [hunk 6]
```diff
@@ -157,152 +161,131 @@ func getBOM(vBom *verrazzanoBom) {
...
+				out, err := exec.Command("kubectl", "get", "pods", "-n", installedNamespace, "-o", "jsonpath=\"{.items[*].spec.containers[*].image}\"").Output()
+				if err != nil {
+					log.Fatal(err)
+				}
+				containerImages += strings.TrimPrefix(strings.TrimSuffix(string(out), `"`), `"`)
+				out, err = exec.Command("kubectl", "get", "pods", "-n", installedNamespace, "-o", "jsonpath=\"{.items[*].spec.initContainers[*].image}\"").Output()
+				if err != nil {
+					log.Fatal(err)
+				}
+				containerImages += strings.TrimPrefix(strings.TrimSuffix(string(out), `"`), `"`)
```
Comment (michael-cico) [inline tools/bom-validator/bom-validator.go:210]: Maybe not for this pass, but would it be useul to track what namespaces an image comes from for reporting purposes?  Might be overkill, what do you think?

File: tools/bom-validator/bom-validator.go [hunk 7]
```diff
@@ -157,152 +161,131 @@ func getBOM(vBom *verrazzanoBom) {
...
+		// Check if the image/tag in the cluster is known to have issues
+		imageWarning, hasKnownIssues := knownImageIssues[nameTag[0]]
+		if hasKnownIssues && vzstring.SliceContainsString(imageWarning.alternateTags, nameTag[1]) {
+			clusterImageWarnings[nameTag[0]] = fmt.Sprintf("Known issue for image %s, found tag %s, expected tag %s message: %s",
+				nameTag[0], nameTag[1], bomImageMap[nameTag[0]], imageWarning.message)
+			continue
+		}
+		// error scenarios,
+		// 1. if cluster's image not found into bom's image map
+		if _, ok := bomImageMap[nameTag[0]]; !ok {
```
[the remark this answers is not in the document]
Comment (desagar) [reply tools/bom-validator/bom-validator.go:279]: I'm a bit confused by the difference between bomImageMap and bomContainerMap. What does a true/false in bomContainerMap mean? Is it the same as bomImageMap[imagename] contains image tag?

File: tools/bom-validator/bom-validator.go (hunk 5 above)
Comment (absBitm) [reply tools/bom-validator/bom-validator.go]: corrected at https://github.com/verrazzano/verrazzano/pull/3275/commits/d63aa28b65acae3ae82ed6049330cf521b9d34ce

File: tools/bom-validator/bom-validator.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (absBitm) [reply tools/bom-validator/bom-validator.go:282]: yes, 
if we want "quick lookup" & "quick report bom tags" both during tags mismatch error case.

File: tools/bom-validator/bom-validator.go [hunk 8]
```diff
@@ -157,152 +161,131 @@ func getBOM(vBom *verrazzanoBom) {
...
+	return strings.Split(strings.TrimSpace(string(out)), "\n")
 }
 
-// ignoreSubComponent - checks to see if a particular subcomponent is to be ignored
-func ignoreSubComponent(name string) bool {
-	for _, subComp := range ignoreSubComponents {
-		if subComp == name {
-			return true
+// Populate a HashMap with all the container & initContainer images found in the cluster
+// Send Cluster's Images Hashmap for BOM Validations
```
[the remark this answers is not in the document]
Comment (absBitm) [reply tools/bom-validator/bom-validator.go]: corrected at https://github.com/verrazzano/verrazzano/commit/d63aa28b65acae3ae82ed6049330cf521b9d34ce

File: tools/bom-validator/bom-validator.go [hunk 9]
```diff
@@ -157,152 +161,131 @@ func getBOM(vBom *verrazzanoBom) {
...
-
-// Validate the images in the BOM against the images found in the cluster
-// It can be
-//    Valid, Tags match
-//    Not Found, OK based on profile
-//    InValid, image tags between BOM and cluster image do not match
-func validateBOM(vBom *verrazzanoBom, clusterImageMap map[string][10]string, imagesNotFound map[string]string,
-	imageTagErrors map[string]imageError, imageWarnings map[string]string) bool {
-	var errorsFound = false
+// Populate bom images and containers into Hashmap
```
[the remark this answers is not in the document]
Comment (absBitm) [reply tools/bom-validator/bom-validator.go]: corrected at https://github.com/verrazzano/verrazzano/commit/d63aa28b65acae3ae82ed6049330cf521b9d34ce

File: tools/bom-validator/bom-validator.go (hunk 4 above)
Comment (absBitm) [reply tools/bom-validator/bom-validator.go]: corrected at https://github.com/verrazzano/verrazzano/commit/d63aa28b65acae3ae82ed6049330cf521b9d34ce

File: tools/bom-validator/bom-validator.go (hunk 3 above)
Comment (absBitm) [reply tools/bom-validator/bom-validator.go]: corrected at https://github.com/verrazzano/verrazzano/commit/d63aa28b65acae3ae82ed6049330cf521b9d34ce

File: tools/bom-validator/bom-validator.go (hunk 6 above)
Comment (absBitm) [reply tools/bom-validator/bom-validator.go:213]: +noted for next development version

**Review by michael-cico — APPROVED** [review]

File: tools/bom-validator/bom-validator.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (desagar) [reply tools/bom-validator/bom-validator.go:282]: I’m not sure why bomImageMap has to have multiple values. 

Also, it looks like you are not using the array of values in bomImageMap[<imagename] except to check whether it exists.

[the remark this answers is not in the document]
Comment (absBitm) [reply tools/bom-validator/bom-validator.go:282]: we are using the array of values in bomImageMap[<imagename],
line-291 : clusterImageTagErrors[nameTag[0]] = imageError{nameTag[1], bomImageMap[nameTag[0]]}

we need both image tag of cluster & image tags of bom for reporting error info.
eg. Check failed! Image Name =  shell , Tag from Cluster =  v0.1.6 Tags from Bom =  [v0.1.16]

[the remark this answers is not in the document]
Comment (absBitm) [reply tools/bom-validator/bom-validator.go:282]: we are using the array of values in bomImageMap[<imagename],
we need both image tag of cluster & image tags of bom for reporting error info.

clusterImageTagErrors[nameTag[0]] = imageError{nameTag[1], bomImageMap[nameTag[0]]}
eg. Check failed! Image Name =  shell , Tag from Cluster =  v0.1.6 Tags from Bom =  [v0.1.16]

**Review by michael-cico — APPROVED** [review]

**Pull request merged**